packages/graph/hash_graph/lib/graph/src/store/query/filter.rs

```diff
@@ -0,0 +1,327 @@
+    fmt::{Debug, Formatter},
+};
+
+use serde::Deserialize;
+
+use crate::store::query::{Expression, Literal, Path, QueryRecord};
+
+/// A set of conditions used for queries.
+#[derive(Deserialize)]
+#[serde(
+    rename_all = "camelCase",
+    bound = "'de: 'q, T::Path<'q>: Deserialize<'de>"
+)]
+pub enum Filter<'q, T: QueryRecord> {
+    All(Vec<Self>),
+    Any(Vec<Self>),
+    Not(Box<Self>),
+    Equal(Option<FilterValue<'q, T>>, Option<FilterValue<'q, T>>),
+    NotEqual(Option<FilterValue<'q, T>>, Option<FilterValue<'q, T>>),
+}
+
+// TODO: Derive traits when bounds are generated correctly
+//   see https://github.com/rust-lang/rust/issues/26925
+impl<'q, T> Debug for Filter<'q, T>
+where
+    T: QueryRecord<Path<'q>: Debug>,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::All(filters) => f.debug_tuple("All").field(filters).finish(),
+            Self::Any(filters) => f.debug_tuple("Any").field(filters).finish(),
+            Self::Not(filter) => f.debug_tuple("Not").field(filter).finish(),
+            Self::Equal(lhs, rhs) => f.debug_tuple("Equal").field(lhs).field(rhs).finish(),
+            Self::NotEqual(lhs, rhs) => f.debug_tuple("NotEqual").field(lhs).field(rhs).finish(),
+        }
+    }
+}
+
+// TODO: Derive traits when bounds are generated correctly
+//   see https://github.com/rust-lang/rust/issues/26925
+impl<'q, T> PartialEq for Filter<'q, T>
+where
+    T: QueryRecord<Path<'q>: PartialEq>,
+{
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::All(lhs), Self::All(rhs)) | (Self::Any(lhs), Self::Any(rhs)) => lhs == rhs,
+            (Self::Not(lhs), Self::Not(rhs)) => lhs == rhs,
+            (Self::Equal(lhs_1, lhs_2), Self::Equal(rhs_1, rhs_2))
+            | (Self::NotEqual(lhs_1, lhs_2), Self::NotEqual(rhs_1, rhs_2)) => {
+                lhs_1 == rhs_1 && lhs_2 == rhs_2
+            }
+            _ => false,
+        }
+    }
+}
+
+impl<'q, T: QueryRecord> TryFrom<Expression> for Filter<'q, T> {
+    type Error = <T::Path<'q> as TryFrom<Path>>::Error;
+
+    fn try_from(expression: Expression) -> Result<Self, Self::Error> {
+        Ok(match expression {
+            Expression::Eq(expressions) => match expressions.len() {
+                0 | 1 => unimplemented!(),
+                2 => Self::Equal(
+                    expressions[0].clone().try_into()?,
+                    expressions[1].clone().try_into()?,
+                ),
+                _ => Self::All(
+                    expressions
+                        .windows(2)
+                        .map(|expressions| {
+                            Ok(Self::Equal(
+                                expressions[0].clone().try_into()?,
+                                expressions[1].clone().try_into()?,
+                            ))
+                        })
+                        .collect::<Result<_, _>>()?,
+                ),
+            },
```

Comment:
Does this have the same performance characteristics?
```suggestion
            Expression::Eq(expressions) => match expressions.as_slice() {
                [] | [_] => unimplemented!(),
                [lhs, rhs] => Self::Equal(lhs.clone().try_into()?, rhs.clone().try_into()?),
                _ => Self::All(
                        expressions
                        .windows(2)
                        .map(|expressions| {
                            Ok(Self::Equal(
                                    expressions[0].clone().try_into()?,
                            expressions[1].clone().try_into()?,
                            ))
                        })
                        .collect::<Result<_, _>>()?,
                )
            }
```
We'd get around having those indices that could be errorful (applies for rest of the outer match block as well)